pages/quests/index.tsx

```diff
@@ -8,23 +8,113 @@ import {
+import { useToast } from '@/hooks/useToast'
 import { useMounted } from '@/hooks/useMounted'
 import { useKeywordSubmit } from '@/hooks/quest/useKeywordSubmit'
+import { useUpdateToken } from '@/hooks/quest/useUpdateToken'
 import { useHasNFT } from '@/hooks/useHasNFT'
+import { useTokenIdOf } from '@/hooks/useTokenIdOf'
+import { useTokenURI } from '@/hooks/useTokenURI'
+import { getContractAddress } from '@/utils/contractAddress'
+import { useUpdateTokenMetadata, KamonToken } from '@/hooks/useUpdateTokenMetadata'
 
 const Quests: NextPage = () => {
   const { t } = useTranslation('common')
+  const { toast } = useToast()
   const mounted = useMounted()
   const { connect, connectors } = useConnect()
   const [metaMask] = connectors
   const { data } = useAccount()
-  const { keyword, inputChange, submit, isSubmitting } = useKeywordSubmit()
+  const { keyword, inputChange, submit, isSubmitting, keywordSubmitSucceeded } = useKeywordSubmit()
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [updateTxLaunched, setUpdateTxLaunched] = useState<boolean>()
+
+  const { updateTokenMetadata, updateTokenMetadataIsSubmitting } = useUpdateTokenMetadata()
+  const { updateToken, updateTokenIsSubmitting } = useUpdateToken(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+
+  useEffect(() => {
+    if (tokenURI) {
+      const fetchData = async () => {
+        const pinataRequest = await fetch(tokenURI.toString())
+        const responseJson = await pinataRequest.json()
+        setTokenJSON(responseJson)
+      }
+      fetchData()
+    }
+  }, [tokenURI])
+
+  const updateTokenMetadataWrapper = async () => {
+    if (data?.address == undefined || tokenJSON == undefined) return
+    const userAddress: string = data?.address
+    toast({
+      title: t('QUEST.TOAST.GENERATING.TITLE'),
+      description: t('QUEST.TOAST.GENERATING.DESCRIPTION'),
+      status: 'info'
+    })
+    const updateTokenMetadataRet: string = await updateTokenMetadata(tokenJSON, userAddress)
+    if (updateTokenMetadataRet.indexOf('Error') === 0) {
+      toast({
+        title: updateTokenMetadataRet,
+        description: t('QUEST.TOAST.UPDATETOKENMETADATA_ERROR.DESCRIPTION'),
+        status: 'error'
+      })
+      return
+    }
+    const finalTokenUri = updateTokenMetadataRet
+
+    setUpdateTxLaunched(false)
+    const theTokenId = tokenIdOf? tokenIdOf: BigInt(0)
+    if(theTokenId == BigInt(0)) { return }
+    setTokenId(BigInt(parseInt(theTokenId.toString())))
+    setFinalTokenUri(finalTokenUri)
+  }
+
+  useEffect(() => {
```

Comment:
also, this one should be in diffrent file too

Reply:
This is really short as of https://github.com/henkaku-center/omise-interface/pull/80/commits/cccee0122a2ab3fd2f730acb61fa20263acf3bfa

```ts
  useEffect(() => {
    if (!finalTokenUri || updateTxLaunched == true) return
    setUpdateTxLaunched(true) // Avoids concurrent transactions
    update()
  }, [finalTokenUri, update, toast, t])
```

We may not need to move this out of `index.tsx` unless we restore those toasts in this PR.